pantheon/src/main/java/tech/pegasys/pantheon/controller/PantheonControllerBuilder.java

```diff
@@ -0,0 +1,308 @@
+  protected Integer networkId;
+  protected MiningParameters miningParameters;
+  protected MetricsSystem metricsSystem;
+  protected PrivacyParameters privacyParameters;
+  protected Path dataDirectory;
+  protected Clock clock;
+  protected Integer maxPendingTransactions;
+  protected KeyPair nodeKeys;
+  private StorageProvider storageProvider;
+  private final List<Runnable> shutdownActions = new ArrayList<>();
+  private RocksDbConfiguration rocksdDbConfiguration;
+
+  public PantheonControllerBuilder<C> rocksdDbConfiguration(
+      final RocksDbConfiguration rocksDbConfiguration) {
+    this.rocksdDbConfiguration = rocksDbConfiguration;
+    return this;
+  }
+
+  public PantheonControllerBuilder<C> storageProvider(final StorageProvider storageProvider) {
+    this.storageProvider = storageProvider;
+    return this;
+  }
+
+  public PantheonControllerBuilder<C> genesisConfigFile(final GenesisConfigFile genesisConfig) {
+    this.genesisConfig = genesisConfig;
+    return this;
+  }
+
+  public PantheonControllerBuilder<C> synchronizerConfiguration(
+      final SynchronizerConfiguration synchronizerConfig) {
+    this.syncConfig = synchronizerConfig;
+    return this;
+  }
+
+  public PantheonControllerBuilder<C> ethereumWireProtocolConfiguration(
+      final EthereumWireProtocolConfiguration ethereumWireProtocolConfiguration) {
+    this.ethereumWireProtocolConfiguration = ethereumWireProtocolConfiguration;
+    return this;
+  }
+
+  public PantheonControllerBuilder<C> networkId(final int networkId) {
+    this.networkId = networkId;
+    return this;
+  }
+
+  public PantheonControllerBuilder<C> miningParameters(final MiningParameters miningParameters) {
+    this.miningParameters = miningParameters;
+    return this;
+  }
+
+  public PantheonControllerBuilder<C> nodePrivateKeyFile(final File nodePrivateKeyFile)
+      throws IOException {
+    this.nodeKeys = loadKeyPair(nodePrivateKeyFile);
+    return this;
+  }
+
+  public PantheonControllerBuilder<C> nodeKeys(final KeyPair nodeKeys) {
+    this.nodeKeys = nodeKeys;
+    return this;
+  }
+
+  public PantheonControllerBuilder<C> metricsSystem(final MetricsSystem metricsSystem) {
+    this.metricsSystem = metricsSystem;
+    return this;
+  }
+
+  public PantheonControllerBuilder<C> privacyParameters(final PrivacyParameters privacyParameters) {
+    this.privacyParameters = privacyParameters;
+    return this;
+  }
+
+  public PantheonControllerBuilder<C> dataDirectory(final Path dataDirectory) {
+    this.dataDirectory = dataDirectory;
+    return this;
+  }
+
+  public PantheonControllerBuilder<C> clock(final Clock clock) {
+    this.clock = clock;
+    return this;
+  }
+
+  public PantheonControllerBuilder<C> maxPendingTransactions(final int maxPendingTransactions) {
+    this.maxPendingTransactions = maxPendingTransactions;
+    return this;
+  }
+
+  public PantheonController<C> build() throws IOException {
```

Comment:
What do you think about moving this class inside `PantheonController` and calling it `PantheonController.Builder`?  It would be more consistent with our other builders.  What complicates it is the fact that the Builder class is subclassed by the other controllers. But to typical users that would be transparent.

Either way, your call.

Reply:
Done.